src/vmm/src/vstate/vm.rs

```diff
@@ -154,24 +200,65 @@ impl Vm {
         })
     }
 
+    fn create_userfault_channels(
+        &self,
+        secret_free: bool,
+    ) -> Result<(Option<UserfaultChannel>, Option<UserfaultChannel>), std::io::Error> {
+        if secret_free {
+            let (receiver_vcpu_to_vm, sender_vcpu_to_vm) = pipe2(libc::O_NONBLOCK)?;
+            let (receiver_vm_to_vcpu, sender_vm_to_vcpu) = pipe2(0)?;
+            Ok((
+                Some(UserfaultChannel {
+                    sender: sender_vcpu_to_vm,
+                    receiver: receiver_vm_to_vcpu,
+                }),
+                Some(UserfaultChannel {
+                    sender: sender_vm_to_vcpu,
+                    receiver: receiver_vcpu_to_vm,
+                }),
+            ))
+        } else {
+            Ok((None, None))
+        }
+    }
```

Comment:
Alright, so I haven't completely digested this part yet, but two questions:

- Why do we need a pipe to send stuff from the vmm thread to the vcpu thread? The vcpu is already paused here, so why can't we reuse the existing rust-channel that we use for sending the resume command?
- Why pipes in the first place? We only need a single eventfd for the vcpu to kick the vmm thread (could even be the same eventfd for all vcpus). The data can then also be sent via a rust-channel (which iirc are completely in userspace)